### ISSUE TITLE
Conservation of quantities

### ISSUE THREAD

**Issue by Mathadon** [iea-annex60/modelica-annex60#247]
We currently do not have (many) unit tests that check if ```MixingVolumes``` conserve mass and energy. I'll make some suggestions that we can have a look at.

**Comment by Mathadon** [comment]
Can we have a look at 
```Annex60.Fluid.MixingVolumes.Validation.TracesubstanceConservation```
and 
```Annex60.Fluid.MixingVolumes.Validation.MoistureMixingConservation```
in https://github.com/Mathadon/modelica-annex60/tree/issue247_conservationTests

The first test fails because ```Annex60.Fluid.Sensors.TraceSubstancesTwoPort``` seems to contain a bug:
```
CMed = s*inStream(port_b.C_outflow);
```
should be
```
CMed = s*inStream(port_a.C_outflow);
```
After fixing this the model still fails. This is because the mass flow rate calculation in ```Annex60.Fluid.Sensors.TraceSubstancesTwoPort``` does not take into account that a part of ```m_flow``` is water vapour. Adding water vapour therefore also increases the mass flow rate of the CO2.

The second test fails because stream connector mixing points assume that ```m_flow``` in a fluid connector is the dry air mass flow rate. 

Both problems can be fixed by setting 
```
  if massDynamics == Modelica.Fluid.Types.Dynamics.SteadyState then
    0 = mb_flow;
  else
    der(m) = mb_flow;
  end if;
``` 
in
```
Annex60.Fluid.Interfaces.ConservationEquation
```
and 
```
port_a.m_flow + port_b.m_flow = 0;
```
in
```
Annex60.Fluid.Interfaces.StaticTwoPortConservationEquation
```. 
This also relates to issue #234 .

**Comment by Mathadon** [comment]
I have not yet implemented the bug fixes so we can test this first and then decide on how to solve it.

**Comment by mwetter** [comment]
@Mathadon 
I slightly revised the models on `issue247_conservationTests` of the `Annex60` repository.
The test setups are good, but there also need to be tests that check the implementation of the dynamic balance. 

Can you please implement how you think the mass balance should be treated, add `.mos` scripts for the new examples and clean up the examples (such as naming convention. expanding the drawing area so the model fits into it etc).

There will also need to be a documentation that if moisture is added, then the trace substance concentration does not change as it is mass fraction per unit mass of dry air and hence adding moisture does not affect the mass flow rate `m_flow`.

**Comment by Mathadon** [comment]
Working on it. Still need to add unit tests. Does it make sense to make an 'empty' unit test without plots? This would reduce overhead since we have already have ```assert``` statements that should verify the implementation.

**Comment by mwetter** [comment]
Adding results in a plot is also instructive for a user (and I am not sure anymore if the regression tests will issue a warning if no results are compared).

**Comment by Mathadon** [comment]
The proposed changes lead to a change in convention: ```m_flow``` becomes the mass flow rate of the dry air and ```Xi``` refers to this mass flow rate for determining the vapour mass: ```mXi = Xi*m_flow```. I think this change needs to be propagated in functions like ```Annex60.Utilities.Psychrometrics.Functions.phi_pTX``` as well, which contains:
```
phi :=p/saturationPressure(T)*X_w/(X_w +
    Annex60.Utilities.Psychrometrics.Constants.k_mair*(1-X_w));
```

Would the 'correct' implementation be:
```
phi :=p/saturationPressure(T)*X_w
```
?

**Comment by Mathadon** [comment]
The proposed changes lead to a change in convention: ```m_flow``` is the mass flow rate of the _dry_ air and ```Xi``` (=m_water/m_dry) refers to this mass flow rate for determining the vapour mass.  Furthermore ```Annex60.Fluid.Sensors.RelativeHumidity``` uses ```Xi``` as an input. This sensor contains following equation:
```
  phi = Annex60.Utilities.Psychrometrics.Functions.phi_pTX(
    p=port.p,
    T=T,
    X_w=Xi[1]);
```
meaning that ```X_w``` in functions such as ```Annex60.Utilities.Psychrometrics.Functions.phi_pTX``` should also be defined as the mass fraction compared to dry air, which is currently not the case.
Functions like ```Annex60.Utilities.Psychrometrics.Functions.phi_pTX``` therefore need to be changed since they currently assume that the input variable ```X_w``` is the concentration (m/m_total) relative the the humid air.

**Comment by Mathadon** [comment]
I have added and run unit tests. Many results change, as expected. For one example the number of systems also increased, which is unexpected. We should analyse whether the changed results make sense but maybe it's better if someone else first verifies my implementation changes. 

A summary/motivation of the changes:
1) The way Modelica/Dymola handles mixing of fluid streams implies that ```m_flow``` is the mass of the dry air and that ```Xi``` allows computing the contained water stream as ```m_flow*Xi```. We used a different convention and therefore the ```MixingVolume``` model needs to be adapted in order for mass to be conserved. 
2) The water mass fraction ```X_w``` for psychrometric calculations used to be defined as ```X_w=Xi``` (m_water/m_total). Since the definition of ```Xi``` has changed this equation is no longer valid. There are now two options: 1) set ```X_w=Xi/(1+Xi)``` and keep the current psychrometric implementations or 2) keep ```X_w=Xi``` and change the psychometric implementations. I opted for the second choice since the psychrometric calculations used to often compute ```x_w``` (m_water/m_dry) as ```x_w := X_w/(1 - X_w);``` since it uses ```x_w``` internally and not ```X_w```. The first option would therefore lead to redundant equations since ```Xi=x_w``` would then be converted into ```X_w``` and then back to ```x_w```. 
I have implemented the second option and updated the documentation accordingly, which now refers to ```X_w``` as the mass fraction compared to the _dry_ air and not the _total_ air.

Note that the changed convention can have a large influence on unit test results where ```Xi``` is an input, since the same numerical value is attributed to a variable which has a changed physical meaning.

**Comment by Mathadon** [comment]
I just realised that we probably need to update ```Annex60.Media.Air.specificEnthalpy``` too, which should be changed from
```
h := (state.T - reference_T)*dryair.cp * (1 - state.X[Water]) +
       ((state.T-reference_T) * steam.cp + h_fg) * state.X[Water];
```
into 
```
h := (state.T - reference_T)*dryair.cp * (1 - state.X[Water])/(1+state.X[Water]) +
       ((state.T-reference_T) * steam.cp + h_fg) * state.X[Water]/(1+state.X[Water]);
```
we should analyse if this does not lead to more systems of non-linear equations.

**Comment by mwetter** [comment]
@Mathadon :
The changes you proposed are in my view confusing as now, the psychometric functions use capital `X` for mass per kg dry air, whereas in Modelica, `X` is mass per kg total air. See also `Annex60.UsersGuide`, in which the convention is that `X` is per kg total air, and `x` is per kg dry air. Using `X` as mass per kg dry air would lead to definitions in some parts of the library that are different from what the media functions and also the MSL are using.
I think this change in definition is not needed. It looks like originally, it appeared that your new definition of `X` leads to fewer operations(?), but in view of the function `h` above, I don't think this is the case.

I added a new branch `issue247_conservationTests_Volume` in which I propose an alternate implementation that only affects the convervation equations of the mixing volume. I thought this is what you proposed to do in your second comment of this issue.
With this implementation, `m_flow` is still the mass flow rate of moist air and `Xi` mixes correctly when multiple ports are connected. If `Xi` were to be defined as mass fraction per kg dry air, and `m_flow` were the dry air mass flow rate, then mixing in ports would be incorrect. 

The approximation that is done in the implementation on `issue247_conservationTests_Volume` only affects the MixingVolume, and advanced users could change this by setting the parameter `approximateMoistureBalance = false`.

Please let me know what you think or whether you like to discuss by phone which may be easiest.

**Comment by Mathadon** [comment]
I looked into this further and I agree that we should probably stick with your implementation. Below is a more detailed analysis for future reference.


When considering the mixing of 2 fluid streams entering in ```port_a``` and 1 exiting stream we get the following Modelica code (if ```Advanced.InStreamRegularizationMode =0```, which avoids further clutter in the code).
```
port_a.Xi_outflow[1] := (max(boundary1.m_flow, 1E-10)*boundary1.X[1]+max(
    boundary.m_flow, 1E-10)*boundary.X[1]+max( -bou.ports[1].m_flow, 1E-10)*
    bou.X[1])/(max(boundary1.m_flow, 1E-10)+max(boundary.m_flow, 1E-10)+max( -
    bou.ports[1].m_flow, 1E-10));
```

This can be simplified into:
	
```
port_a.Xi_outflow[1] := (boundary1.m_flow*boundary1.X[1]+
    boundary.m_flow*boundary.X[1]+1E-10*bou.X[1])/
	(1E-10 + 1E-10 - bou.ports[1].m_flow);
```	
or
```
port_a.Xi_outflow[1]*bou.ports[1].m_flow + boundary1.m_flow*boundary1.X[1]+ boundary.m_flow*boundary.X[1])=0;
```
.

	
Therefore the conserved quantity in this equation is ```m_flow*X[1]```. Since water vapor mass is what needs to be conserved then ```m_flow*X[1]``` needs to be equal to the water vapor mass. I see two options of how this can be achieved:

1) 
m_flow is the total wet air mass (m_wet) and Xi is the water vapor mass fraction compared to this wet mass: (m_Water/m_wet) such that the product of m_flow and Xi equals m_water.
2) 
m_flow is the dry air mass (m_dry) and Xi is the water vapor mass fraction compared to the dry air mass (m_water/m_dry) such that the product of m_flow and Xi equals m_water. Based on that analysis I disagree with
>  If Xi were to be defined as mass fraction per kg dry air, and m_flow were the dry air mass flow rate, then mixing in ports would be incorrect.
Both options seem to be possible?

The current implementation is 1). I thought this was wrong since the verification test (```Annex60.Fluid.MixingVolumes.Validation.MoistureMixingConservation```) returned an error and I thought that we therefore would have to switch to option 2). After further analysis this however does not seem to be the case since the error is caused by ```Annex60.Fluid.Interfaces.StaticTwoPortConservationEquation``` not taking into account ```mWat_flow``` in equation
```
port_b.Xi_outflow = inStream(port_a.Xi_outflow) + mXi_flow * m_flowInv;
```
this should become something like
```
port_b.Xi_outflow = (inStream(port_a.Xi_outflow)*port_a.m_flow + mXi_flow) * Annex60.Utilities.Math.Functions.inverseXRegularized(x=-port_b.m_flow, delta=m_flow_small/1E3);
```
You already indicated in the code that this is an approximation but in my opinion the approximation should not be hard-coded since conservation of mass is quite fundamental. Fixing this equation would fix conservation of mass. 

However changing to option 1) (which is the implementation I made) would also 'fix' issue #234 without requiring an approximation. I was unaware of the convention regarding X_w and x_w. This would then need to be changed, making some models non backwards compatible since some model inputs/outputs need to be renamed. I do not think the changed implementation would lead to more non-linear systems in typical applications. I reran the unit tests and they seem to confirm this. The downside is that MSL Media would be incompatible with Annex 60 since they seem to have chosen Xi to refer to moist air.

To summarise:
Option 1) : wet air as reference
+ backwards compatible
+ less operations in enthalpy calculation
+ compatible with MSL
- more operations in psychrometry functions and in StaticTwoPortConservationEquation
- need parameter for fixing #234 , which causes a violation of conservation of mass


Option 2) : dry air as reference
+ fixes #234 naturally
+ less operations in psychrometry functions and in StaticTwoPortConservationEquation
- more operations in enthalpy calculation
- not backwards compatible
- not compatible with MSL

Taking all of that into consideration then I propose to continue with the implementation you proposed, but to ‘fix’ the equation in ```Annex60.Fluid.Interfaces.StaticTwoPortConservationEquation```. It's probably also best to make ```approximateMoistureBalance``` a parameter with ```annotation(Evaluate=true)``` and to propagate it into ```MixingVolume```.

**Comment by mwetter** [comment]
@Mathadon 
Thanks for the explanation. So let's proceed with option 1. I can look into the reformulation of `port_b.Xi_outflow = (inStream(...` next week.

The rationale for making `approximateMoistureBalance` a constant was so that it does not show up in the GUI. I could go either way, but was selecting a parameter as I am not sure that changing it to `false` will be used by anyone other than developers for testing and profiling the code. For backward compatibility, it will be easier to change a constant to a parameter if there will be a need than the other way around as the latter can make models invalid.

**Comment by Mathadon** [comment]
okay! I made a pull request with a suggestion for the implementation of Xi_outflow in the meantime.

**Comment by Mathadon** [comment]
I updated code in https://github.com/iea-annex60/modelica-annex60/tree/issue247_conservationTestsv2

However there are some warnings in the new models since we need some protected variables in dynBal. This will cause conflicts with pedantic mode. @mwetter is there a way to unprotect `dynBal` in the example somehow?

**Comment by mwetter** [comment]
@Mathadon : The Modelica language does not allow you to make instances public once they were declared as protected. Hence, the examples will need to be changed so that they conform to the language standard.

**Comment by Mathadon** [comment]
The most practical solution is to make some variables in `dynBal` public. Is this an option?

**Comment by mwetter** [comment]
But if they are public, then the output file will increase. I prefer to
keep variables protected unless they are indeed of interest to the user.

**Comment by Mathadon** [comment]
Could discuss this during today's meeting!

**Comment by Mathadon** [comment]
I should try the following: create a public output U of the protected model and then assign the value in the higher level model.

**Comment by Mathadon** [comment]
I created outputs `m`, `mC`, `mXi` and `U` for `MixingVolume`. In my opinion it makes sense to output these variables since these quantities is all what the model is about.

**Comment by Mathadon** [comment]
Todo:

- check documentation
- check quantities and units
- explain unit test difference

**Comment by Mathadon** [comment]
Fixed in #436
